### ISSUE TITLE
Arma: Cold War Assault - Artifacts with EAX enabled

### ISSUE THREAD

**Issue by Hiradur** [kcat/dsoal#16]
In Arma: Cold War Assault with DSOAL and the `-dsound` command line argument turning on EAX causes artifacts on footsteps originating from the player character. The strength of these artifacts depend on the EAX environment of the player character. Toggling proning may also cause artifacts but I'm unsure if that is what it's supposed to sound like.
Not sure what is happening but it sounds a bit like the reverb effect is much too strong for these sound effects. Sound effects not originating from the player character seem to play back with normal reverb and no artifacts.

A good place to test this is the beach at the beginning of the single mission "Ambush". Right from the start the artifacts should be audible when walking.

On YouTube there is a video of Operation Flashpoint [1] (Arma: Cold War Assault is a re-release of that game) where EAX is apparently enabled on a PC with Windows 2000 and SoundBlaster Audigy 2 ZS. There the artifacts are not audible (I tested the same mission) so I believe the fault is not in the game itself.

Tested on Ubuntu 18.04 with Arma: Cold War Assault version 1.99 from GOG running on Lutris using the `lutris-5.7-5-x86_64` Wine build. Judging from the DSOAL Logfile generated with `DSOAL_LOGLEVEL=3` I believe the game is using EAX 2.

I attached a short recording [2] of me walking at the beach of the mission "Ambush". The first 10 seconds are with EAX enabled, the rest is with EAX disabled.

[1] https://youtu.be/kngkmKleUo8?t=334
[2] [arma_eax_bug_dsoal.zip](https://github.com/kcat/dsoal/files/4932149/arma_eax_bug_dsoal.zip)

**Comment by Hiradur** [comment]
The recording linked in the inital post contained only audio. I made a video recording that hopefully better demonstrates the issue. Corresponding log files are also in the attached archive. Everything was recorded at the beginning of the 5th single mission called "Heavy Metal".
As you can hear in the video, the artifacts sometimes disappear. I don't really know what's causing this.

[arma_eax_bug_video.zip](https://github.com/kcat/dsoal/files/6719956/arma_eax_bug_video.zip)

**Comment by ThreeDeeJay** [comment]
I was also able to replicate the bug on the Steam version with a [recent DSOAL + OpenAL Soft](https://airtable.com/shrtTsUtiAJQ7wFGI/tblNOTdmp5nHXfFGU/viwjDPuKOjBnIQunv/recCGP4GXqgRnnTDU): https://i.imgur.com/WAHXRfr.mp4
Tho like you mention, it seems to be a DSOAL-specific issue because [Creative ALchemy with an X-Fi sound card](https://airtable.com/shrtTsUtiAJQ7wFGI/tblNOTdmp5nHXfFGU/viwjDPuKOjBnIQunv/recJecfbW7hTgAuXB) under the same conditions just works™: https://i.imgur.com/52jJXNA.mp4

**Comment by Hiradur** [comment]
Issue is still present with OpenAL Soft 1.22.2 official release and DSOAL commit a16c19f.

**Comment by kcat** [comment]
It almost sounds like there's two instances of each sound, creating a slight "phasy" effect, and sometimes the reverb becoming very strong temporarily. I don't see why this would be happening, though. The log shows the only EAX calls being made are first to query support, then very occasionally switches back and forth between the City and Plain presets. There is an oddity that every time it switches presets, it then sets the environment size to 0 when the range is 1 to 100. Older DSOAL I think would clamp the range (which could've accounted for it, since reducing the size to the minimum would increase the reverb gain and reduce the delays), but newer DSOAL/OpenAL Soft should error when given a size of 0 and leave the presets unchanged.

**Comment by Hiradur** [comment]
> It almost sounds like there's two instances of each sound, creating a slight "phasy" effect, and sometimes the reverb becoming very strong temporarily.

While the recording indeed sounds a bit _phasy_ (and I can't notice that _phasiness_ in the recoding linked by @ThreeDeeJay), the issue I was referring to is the loud noise coming from the left speaker while moving. It can be heard in the video recording I uploaded as part of the zip file in https://github.com/kcat/dsoal/issues/16#issuecomment-868982808.

**Comment by kcat** [comment]
> While the recording indeed sounds a bit _phasy_ (and I can't notice that _phasiness_ [in the X-Fi recoding linked by](https://i.imgur.com/52jJXNA.mp4) @ThreeDeeJay), the issue I was referring to is the loud noise coming from the left speaker while moving.

Yes, my thinking was these may be related. It seems like there's two instances of the sound timed very close to each other, which could've been the result of the game setting a reverb preset followed by setting the environment size to 0. On older DSOAL versions, this would result in the environment size being clamped to the minimum value of 1. By reducing the environment size to such a small value from the preset's original value, it would modify the initial reflections and late reverb delays to very small values, along with increasing the gains (and shortening the decay times). This would result in a sound being heard more clearly multiple times very closely in time, creating that phasiness.

Since the game occasionally swaps between two reverb presets, City and Plain, it made me think the "loud noise coming from the left speaker" is one of these reverbs having its environment size reduced to minimum and becoming much louder than the other, which goes away when it switches back to the other preset.

If that were the case, though, then newer DSOAL/OpenAL Soft should behave differently. With the native EAX functions, setting the environment size to 0 should error and leave the preset unchanged. The phasiness and loud noise should've stopped happening since it would be using the presets as-is without the size alteration. If it didn't change anything though, then that would mean the environment size change isn't causing these artifacts, or some other change is still happening.

It would be helpful if you're able to make updated logs with the newer versions, so I can check that DSOAL and OpenAL Soft are behaving as hypothesized.

**Comment by Hiradur** [comment]
Sorry, I just realized that the doubling of sound effects probably was an artifact caused by incorrect settings in OBS for the recording. The phasiness probably was not present in the game at the time I made the recording.

Anyway, the loud noise is still present with OpenAL Soft 1.22.2 and DSOAL commit a16c19f. I attached a new recording, this time with correct OBS settings, hopefully...

[2022-07-03-arma_eax_bug.zip](https://github.com/kcat/dsoal/files/9035508/2022-07-03-arma_eax_bug.zip)

**Comment by kcat** [comment]
This is still using an older DSOAL:
```
0124:trace:dsound:DSShare_Create Found ALC_EXT_EFX
```
The newer version doesn't check for this extension anymore, and should instead be checking for `EAX5.0`, which it isn't. The logs are also showing the old traces for the EAX->EFX wrapper functions that don't exist anymore.

**Comment by Hiradur** [comment]
> This is still using an older DSOAL

You are right and I'm sorry for that. The game apparently is hard-coded to look for `dsound.dll` in the `Windows\System32` directory and I forgot about that. I put the newer build in the game folder but not in `Windows\System32`, where I had an older build still lying around.

Anyway, here is another recording. With the new DSOAL build, it sounds different. The noise is still audible but much less obtrusive. In the second mission called Ambush (not recorded) it's much closer to the recordings of real hardware [1](https://youtu.be/kngkmKleUo8?t=334) and [2](https://i.imgur.com/52jJXNA.mp4). In fact, I can hear some low level rumbling there too while they are walking. However, in the 5th single mission called Heavy Metal it's more pronounced and I'm wondering whether it's the game's fault or not. @ThreeDeeJay Should you have real hardware available for making comparisons, could you test the 5th mission and post your impressions or a recording?

[2022-07-04-arma_eax_bug.zip](https://github.com/kcat/dsoal/files/9042033/2022-07-04-arma_eax_bug.zip)

**Comment by kcat** [comment]
DSOAL's reverb seems to actually be quieter compared to ALchemy and real hardware (there's more apparent reverb during the reload sound, for example), and the heart pumping(?) sound when running is less pronounced with DSOAL.

**Comment by Hiradur** [comment]
Revisting this issue, I´d like to summarize my observations from listening:
* The artifact I was talking about is still present with current builds of OpenAL Soft (46df54724403d9c25e1cded8a417d72ae4db5434) and DSOAL (14a8ca579b20137e10ced47987a34ed8de9f8da6)
* The artifact appears when enabling EAX while using DSOAL with OpenAL Soft
* One particular sound is causing this "artifact". It is way too loud with EAX enabled.
  * This sound is tied to the player character and moves with them.
  * This sound is very close to the listener but not listener-local, so I think it is a 3D sound.
    * You can confirm this by switching from first-person to third-person perspective, where the sound becomes (fully?) inaudible
  * With HRTF, you can hear it being on the left side but close to the ear.
  * The sound only appears if the player is physically moving (no matter if they walk, run or crouch)
    * I _think_ movement speed slightly affects the intensity of the sound
    * I think this sound might represent either air movement in the player's ears or friction in the player's clothes while moving.
  * The sound does not appear if the player is driving in a vehicle  
  * This sound is also audible without EAX enabled but it is _very_ quiet (best heard while walking)
  * The appearence of the "artifact" is influenced by EAX environments. You can clearly hear this at the spawn point of single mission number 5 "Heavy Metal", where you basically spawn on the border of two EAX environments. Here, you can quickly compare between the two.
* Additionally, there are breathing and heart-pumping sounds while running, which are also tied to the player character. They don't suffer from the same issue.
  * While moving forward, they _feel_ listener-local but aren't: when walking left or right after running, they are either on the left or right side of the listener. They are also rather close to the ears, yet don't exhibit the same problem as that other sound. I think they are further away from the listener in this case than the problematic sound, though. When moving forward they appear to be centered on the listener, however.
* The issue is neither caused by HRTF nor Ambisonics rendering. The "artifact" is audible with headphones as well as surround speakers.

Here are the traces using fresh build using the commits mentioned above:
[arma_dsoal_eax_artifact_202409.zip](https://github.com/user-attachments/files/17002346/arma_dsoal_eax_artifact_202409.zip)

Since only one sound seems to be problematic and it is very close to the listener but not listener-local, I wonder if there are unintended side effects with how OpenAL Soft handles very close sounds and EAX together. However, the heart-pumping and breathing sounds are also 3D sounds close to the listener and don't show the same problem :thinking: .

**Comment by kcat** [comment]
> Here are the traces using fresh build using the commits mentioned above:
> [arma_dsoal_eax_artifact_202409.zip](https://github.com/user-attachments/files/17002346/arma_dsoal_eax_artifact_202409.zip)

This seems to be using the latest master of DSOAL, not the c++-rewrite with that commit (it's showing C function names and using old extensions).

> Since only one sound seems to be problematic and it is very close to the listener but not listener-local, I wonder if there are unintended side effects with how OpenAL Soft handles very close sounds and EAX together. However, the heart-pumping and breathing sounds are also 3D sounds close to the listener and don't show the same problem 🤔 .

Unfortunately the game seems to set parameters using `SetAllParameters`, so it doesn't show what values are being set. It's possible DSound has some kind of "deadzone" for nearby sounds, so a sound that's ever-so-slightly to the left should be centered instead if left-panned. Alternatively, maybe there's some issue with how `DS3DMODE_DISABLE` and/or `DSBCAPS_CTRLPAN` is handled if that sound is using that mode.

**Comment by Hiradur** [comment]
> This seems to be using the latest master of DSOAL, not the c++-rewrite with that commit (it's showing C function names and using old extensions).

Right, sorry. I fell into the same trap I already fell into in https://github.com/kcat/dsoal/issues/16#issuecomment-1174367811. Nevertheless, the issue still exists with the `c++-rewrite`. FWIW here are the logs of DSOAL c++-rewrite commit 14a8ca57: [arma_eax_issue_20240918.zip](https://github.com/user-attachments/files/17040346/arma_eax_issue_20240918.zip)


I tried to listen closely [to this recording on original hardware](https://youtu.be/UPrw_biI_WI?t=23), but I was unable to hear the problematic sound in this case. A longer recording of just walking (not running and shooting) around would be helpful. @ThreeDeeJay , could you please check if you can hear the problematic sound toward the left with EAX enabled on real hardware while walking? Might it be centered in this configuration instead?
I suggest that you test using the 5th single mission "heavy metal" since there are two different EAX environments close at the spawn point, which simplifies testing. The border of the EAX environments lies ca. 2m behind the town signs of `Levie` at the spawn point.

**Comment by Hiradur** [comment]
The issue is still present with DSOAL @ 3a19511810a988cfe536007548d1513c9abe81fb and OpenAL Soft @ dd4e07de0fe73d8c0326c4502f63e71da8ef268b.

Not sure if it could be of any help to identify the root cause of the issue, but [the source code of Arma: Cold War Assault Remastered has recently been open sourced](https://github.com/BohemiaInteractive/CWR).
The remaster contains an OpenAL backend with EFX support, which, according to some comments in the code, contains 1:1 ports from the DX8 audio code.

I attached new logs since some of the logged messages are more verbose now compared to 2020.

[arma_cwa_eax_issue_2026.zip](https://github.com/user-attachments/files/29349994/arma_cwa_eax_issue_2026.zip)

**Comment by Hiradur** [comment]
https://github.com/BohemiaInteractive/CWR/issues/15 explains what is going on. It seems to be an issue in the game engine. As a workaround, the author recommends to set

```ini
[reverb]
boost = -12
```

in `alsoft.ini`.

I'm closing this issue since it is not DSOAL's fault.

**Comment by mirh** [comment]
It was literally argued this wasn't happening on the original soundblaster hardware.

**Comment by Hiradur** [comment]
Right, that was my impression from a video. Creative's driver might already have included a fix for this Arma issue.

@ThreeDeeJay, since you uploaded the video, can you tell if you hear that air friction simulating sound with original hardware?

**Comment by mirh** [comment]
Creative might not have had the bug to begin with.
Again, it's not a secret at all that dsoal has many shortcomings (as you even seen the other day with deus ex).

**Comment by Hiradur** [comment]
Maybe. However, in the comments of the video I can see that a USB X-Fi was used. I might remember this incorrectly, but I read some years ago that the USB X-Fi uses software emulation for EAX, which was claimed to not be up to par with the hardware implementation. If this was the case, it couldn't really be seen as a reference either.

**Comment by mirh** [comment]
Not *the* reference, but still one example of no artifacts.
Also, putting aside that the first video you linked is an Audigy 2 ZS on W2000, 3DJ used his [SB1090](https://github.com/kcat/dsoal/issues/36#issuecomment-1399351362) on XP (where it's actually not demonstrated that it sucks).

**Comment by Hiradur** [comment]
Assuming the issue's cause lies outside the game engine, the first step to see if it's caused by DSOAL or OpenAL Soft would be to try the OpenAL backend of ARMA CWR with OpenAL Soft and other OpenAL drivers and see which drivers suffer from the same issue. If the issue was only present with OpenAL Soft but no other AL drivers, OpenAL Soft might cause the issue.
